// @vtex/styleguide/lib/EXPERIMENTAL_TableTree/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Toolbar = require("../EXPERIMENTAL_Table/Toolbar");

var _Toolbar2 = _interopRequireDefault(_Toolbar);

var _DataTable = require("../EXPERIMENTAL_Table/DataTable");

var _DataTable2 = _interopRequireDefault(_DataTable);

var _Tree = require("./Tree");

var _Tree2 = _interopRequireDefault(_Tree);

var _EXPERIMENTAL_Table = require("../EXPERIMENTAL_Table");

var _Pagination = require("../EXPERIMENTAL_Table/Pagination");

var _Pagination2 = _interopRequireDefault(_Pagination);

var _FilterBar = require("../EXPERIMENTAL_Table/FilterBar");

var _FilterBar2 = _interopRequireDefault(_FilterBar);

var _constants = require("../EXPERIMENTAL_useCheckboxTree/constants");

var _Headings = require("../EXPERIMENTAL_Table/DataTable/Headings");

var _Headings2 = _interopRequireDefault(_Headings);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var TableTree = function TableTree(_ref) {
  var children = _ref.children,
      columns = _ref.columns,
      items = _ref.items,
      checkboxes = _ref.checkboxes,
      nodesKey = _ref.nodesKey,
      comparator = _ref.comparator,
      measures = _ref.measures,
      loading = _ref.loading,
      empty = _ref.empty,
      emptyState = _ref.emptyState,
      onRowClick = _ref.onRowClick;

  if (!measures) {
    throw new Error('Provide measures to the TableTree');
  }

  var tableHeight = measures.tableHeight,
      rowHeight = measures.rowHeight,
      selectedDensity = measures.selectedDensity;
  return _react2.default.createElement("div", {
    style: {
      minHeight: tableHeight
    },
    className: "flex flex-column"
  }, children, _react2.default.createElement(_DataTable2.default, {
    loading: loading,
    emptyState: emptyState,
    empty: empty,
    height: tableHeight
  }, _react2.default.createElement("thead", {
    className: "w-100 ph4 truncate overflow-x-hidden c-muted-2 f6"
  }, _react2.default.createElement(_Headings2.default, {
    checkboxes: checkboxes,
    columns: columns
  })), !empty && !loading && _react2.default.createElement("tbody", null, _react2.default.createElement(_Tree2.default, {
    selectedDensity: selectedDensity,
    checkboxes: checkboxes,
    columns: columns,
    items: items,
    onRowClick: onRowClick,
    comparator: comparator,
    nodesKey: nodesKey,
    rowHeight: rowHeight
  }))));
};

TableTree.defaultProps = {
  nodesKey: 'children',
  comparator: _constants.defaultComparatorCurry
};
var treePropTypes = {
  nodesKey: _propTypes2.default.string,
  comparator: _propTypes2.default.func
};

var propTypes = _extends({}, treePropTypes, {}, _EXPERIMENTAL_Table.tablePropTypes);

TableTree.Toolbar = _Toolbar2.default;
TableTree.FilterBar = _FilterBar2.default;
TableTree.Pagination = _Pagination2.default;
TableTree.propTypes = propTypes;
exports.default = TableTree;
