// @vtex/styleguide/lib/EXPERIMENTAL_Table/BulkActions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _constants = require("../constants");

var _Actions = require("./Actions");

var _Actions2 = _interopRequireDefault(_Actions);

var _Tail = require("./Tail");

var _Tail2 = _interopRequireDefault(_Tail);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BULK_ACTIONS_HEIGHT = 56;
var BULK_ACTIONS_TRANSITION = 'height 0.2s ease-in-out, padding 0.2s ease-in-out';

var BulkActions = function BulkActions(_ref) {
  var _ref$active = _ref.active,
      active = _ref$active === void 0 ? false : _ref$active,
      children = _ref.children;
  var positionFixer = _react2.default.Children.count(children) > 1 ? null : _react2.default.createElement("div", {
    className: _constants.ORDER_CLASSNAMES.BULK_CHILD.POSITION_FIXER
  });
  return _react2.default.createElement("div", {
    id: _constants.NAMESPACES.BULK_ACTIONS,
    className: (0, _classnames2.default)('flex flex-row justify-between bg-action-primary c-on-action-primary br3 br--top ph4', {
      pv4: active
    }, _constants.ORDER_CLASSNAMES.BULK),
    style: {
      height: active ? BULK_ACTIONS_HEIGHT : 0,
      overflow: active ? 'auto' : 'hidden',
      transition: BULK_ACTIONS_TRANSITION
    }
  }, children, positionFixer);
};

BulkActions.Actions = _Actions2.default;
BulkActions.Tail = _Tail2.default;
exports.default = BulkActions;
