// @vtex/styleguide/lib/EXPERIMENTAL_Table/DataTable/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _index = require("../../EmptyState/index.js");

var _index2 = _interopRequireDefault(_index);

var _constants = require("../constants");

var _useTableMeasures = require("../hooks/useTableMeasures");

var _Loading = require("./Loading");

var _Loading2 = _interopRequireDefault(_Loading);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataTable = function DataTable(_ref) {
  var children = _ref.children,
      height = _ref.height,
      className = _ref.className,
      Tag = _ref.as,
      empty = _ref.empty,
      loading = _ref.loading,
      emptyState = _ref.emptyState;
  var showLoading = !empty && loading;
  var showEmptyState = empty && emptyState;
  return _react2.default.createElement("div", {
    id: _constants.NAMESPACES.TABLE,
    style: {
      minHeight: height
    },
    className: (0, _classnames2.default)('order-1 mw-100 overflow-x-auto', _constants.ORDER_CLASSNAMES.TABLE)
  }, _react2.default.createElement(Tag, {
    className: "w-100 " + className,
    style: {
      borderSpacing: 0
    }
  }, children), showLoading && _react2.default.createElement(_Loading2.default, {
    height: height - _useTableMeasures.TABLE_HEADER_HEIGHT
  }, typeof loading !== 'boolean' && loading.renderAs && loading.renderAs()), showEmptyState && _react2.default.createElement(_index2.default, {
    title: emptyState.label
  }, emptyState.children));
};

DataTable.defaultProps = {
  as: 'table',
  className: ''
};
exports.default = DataTable;
