// @vtex/styleguide/lib/EXPERIMENTAL_Table/DataTable/Row.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Cell = undefined;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _uuid = require("uuid");

var _uuid2 = _interopRequireDefault(_uuid);

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Row = function Row(_ref) {
  var _ref$as = _ref.as,
      Tag = _ref$as === void 0 ? 'tr' : _ref$as,
      children = _ref.children,
      height = _ref.height,
      onClick = _ref.onClick,
      active = _ref.active;
  var className = (0, _classnames2.default)('w-100 ph4 truncate overflow-x-hidden', {
    'pointer hover-c-link hover-bg-muted-5': onClick,
    'bg-action-secondary': active
  });
  return _react2.default.createElement(Tag, {
    key: _constants.NAMESPACES.ROW + "-" + (0, _uuid2.default)(),
    style: {
      height: height
    },
    onClick: onClick,
    className: className
  }, children);
};

var Cell = exports.Cell = function Cell(_ref2) {
  var id = _ref2.id,
      children = _ref2.children,
      width = _ref2.width,
      onClick = _ref2.onClick,
      _ref2$as = _ref2.as,
      Tag = _ref2$as === void 0 ? 'td' : _ref2$as,
      _ref2$className = _ref2.className,
      className = _ref2$className === void 0 ? '' : _ref2$className;
  var classNames = (0, _classnames2.default)('truncate v-mid ph2 pv0 tl bb b--muted-4', className, {
    'pointer hover-c-link hover-bg-muted-5': onClick
  });
  return _react2.default.createElement(Tag, {
    onClick: onClick,
    id: _constants.NAMESPACES.CELL + "-" + id,
    style: {
      minWidth: width
    },
    className: classNames
  }, children);
};

Row.Cell = Cell;
exports.default = Row;
