// @vtex/styleguide/lib/EXPERIMENTAL_Table/Toolbar/ButtonGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _index = require("../../icon/Download/index.js");

var _index2 = _interopRequireDefault(_index);

var _index3 = require("../../icon/Upload/index.js");

var _index4 = _interopRequireDefault(_index3);

var _Button = require("./Button");

var _Button2 = _interopRequireDefault(_Button);

var _ButtonColumns = require("./ButtonColumns");

var _ButtonColumns2 = _interopRequireDefault(_ButtonColumns);

var _ButtonDensity = require("./ButtonDensity");

var _ButtonDensity2 = _interopRequireDefault(_ButtonDensity);

var _ButtonExtraActions = require("./ButtonExtraActions");

var _ButtonExtraActions2 = _interopRequireDefault(_ButtonExtraActions);

var _ButtonNewLine = require("./ButtonNewLine");

var _ButtonNewLine2 = _interopRequireDefault(_ButtonNewLine);

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var getButton = function getButton(type, props) {
  switch (type) {
    case ButtonType.Columns:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.COLUMNS
        }, _react2.default.createElement(_ButtonColumns2.default, props));
      }

    case ButtonType.Density:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.DENSITY
        }, _react2.default.createElement(_ButtonDensity2.default, props));
      }

    case ButtonType.Download:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.DOWNLOAD
        }, _react2.default.createElement(_Button2.default, _extends({
          id: _constants.NAMESPACES.TOOLBAR.BUTTON_DOWNLOAD,
          icon: _react2.default.createElement(_index2.default, {
            size: _Button.IconSize.Heavy
          })
        }, props)));
      }

    case ButtonType.Upload:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.UPLOAD
        }, _react2.default.createElement(_Button2.default, _extends({
          id: _constants.NAMESPACES.TOOLBAR.BUTTON_UPLOAD,
          icon: _react2.default.createElement(_index4.default, {
            size: _Button.IconSize.Heavy
          })
        }, props)));
      }

    case ButtonType.ExtraActions:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.EXTRA_ACTIONS
        }, _react2.default.createElement(_ButtonExtraActions2.default, props));
      }

    case ButtonType.NewLine:
      {
        return _react2.default.createElement("span", {
          className: _constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP_CHILD.NEWLINE
        }, _react2.default.createElement(_ButtonNewLine2.default, props));
      }

    default:
      {
        return null;
      }
  }
};

var getComponent = function getComponent(type) {
  return function (props) {
    return getButton(type, props);
  };
};

var ButtonGroup = function ButtonGroup(_ref) {
  var children = _ref.children;
  return _react2.default.createElement("div", {
    id: _constants.NAMESPACES.TOOLBAR.BUTTON_GROUP,
    className: (0, _classnames2.default)(_constants.ORDER_CLASSNAMES.TOOLBAR_CHILD.BUTTON_GROUP, 'flex flex-row items-center')
  }, children);
};

var ButtonType;

(function (ButtonType) {
  ButtonType[ButtonType["Columns"] = 0] = "Columns";
  ButtonType[ButtonType["Density"] = 1] = "Density";
  ButtonType[ButtonType["Download"] = 2] = "Download";
  ButtonType[ButtonType["Upload"] = 3] = "Upload";
  ButtonType[ButtonType["ExtraActions"] = 4] = "ExtraActions";
  ButtonType[ButtonType["NewLine"] = 5] = "NewLine";
})(ButtonType || (ButtonType = {}));

ButtonGroup.Columns = getComponent(ButtonType.Columns);
ButtonGroup.Density = getComponent(ButtonType.Density);
ButtonGroup.Download = getComponent(ButtonType.Download);
ButtonGroup.Upload = getComponent(ButtonType.Upload);
ButtonGroup.ExtraActions = getComponent(ButtonType.ExtraActions);
ButtonGroup.NewLine = getComponent(ButtonType.NewLine);
exports.default = ButtonGroup;
