// @vtex/styleguide/lib/EXPERIMENTAL_Table/Toolbar/PopoverMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Alignment = exports.Item = exports.Box = undefined;
exports.default = usePopoverMenu;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _index = require("../../Button/index.js");

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function usePopoverMenu() {
  var _useState = (0, _react.useState)(false),
      boxVisible = _useState[0],
      setBoxVisible = _useState[1];

  var buttonRef = (0, _react.useRef)(null);

  var handleOutsideClick = function handleOutsideClick(e) {
    return buttonRef && buttonRef.current && e.target instanceof Node && !buttonRef.current.contains(e.target) && setBoxVisible(false);
  };

  (0, _react.useLayoutEffect)(function () {
    if (boxVisible) {
      document.addEventListener('mousedown', handleOutsideClick);
    }

    return function () {
      return document.removeEventListener('mousedown', handleOutsideClick);
    };
  }, [boxVisible]);
  var toggleBox = (0, _react.useCallback)(function () {
    setBoxVisible(!boxVisible);
  }, [boxVisible]);
  return {
    boxVisible: boxVisible,
    setBoxVisible: setBoxVisible,
    buttonRef: buttonRef,
    toggleBox: toggleBox
  };
}

var Box = exports.Box = function Box(_ref) {
  var alignMenu = _ref.alignMenu,
      height = _ref.height,
      width = _ref.width,
      noMargin = _ref.noMargin,
      borderClasses = _ref.borderClasses,
      groupActions = _ref.groupActions,
      children = _ref.children;
  var isAlignRight = alignMenu === Alignment.Right;
  return _react2.default.createElement("div", {
    className: "absolute z-999 shadow-4 " + (isAlignRight ? 'right-0' : 'left-0') + " " + (borderClasses || 'b--muted-4 br2 ba') + " " + (noMargin ? '' : 'mt2 mh2'),
    style: {
      width: width
    }
  }, _react2.default.createElement("div", {
    className: "w-100 b2 br2 bg-base"
  }, groupActions && _react2.default.createElement("div", {
    className: "flex inline-flex bb b--muted-4 w-100 justify-center pv4"
  }, groupActions.map(function (action) {
    return _react2.default.createElement("div", {
      className: "mh2",
      key: action.id
    }, _react2.default.createElement(_index2.default, {
      variation: "secondary",
      size: "small",
      onClick: action.onClick
    }, action.label));
  })), _react2.default.createElement("div", {
    className: "overflow-auto",
    style: {
      height: height
    }
  }, children)));
};

var Item = exports.Item = function Item(_ref2) {
  var isSelected = _ref2.isSelected,
      onClick = _ref2.onClick,
      children = _ref2.children;
  return _react2.default.createElement("div", {
    className: "flex justify-between ph6 pv3 " + (isSelected ? 'b--emphasis' : 'b--transparent') + " pointer hover-bg-muted-5 bl bw1",
    onClick: onClick
  }, _react2.default.createElement("span", {
    className: "w-100 flex justify-between " + (isSelected ? 'fw5' : '')
  }, children));
};

var Alignment = exports.Alignment = undefined;

(function (Alignment) {
  Alignment["Left"] = "left";
  Alignment["Right"] = "right";
})(Alignment || (exports.Alignment = Alignment = {}));
