// @vtex/styleguide/lib/EXPERIMENTAL_Table/hooks/useTableVisibility.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useTableVisibility;

var _react = require("react");

function useTableVisibility(_ref) {
  var columns = _ref.columns,
      _ref$hiddenColumns = _ref.hiddenColumns,
      initHiddenColumns = _ref$hiddenColumns === void 0 ? [] : _ref$hiddenColumns;

  var _useState = (0, _react.useState)(initHiddenColumns),
      hiddenColumns = _useState[0],
      setHiddenColumns = _useState[1];

  var visibleColumns = (0, _react.useMemo)(function () {
    var reducer = function reducer(acc, col) {
      return hiddenColumns.includes(col.id) ? acc : [].concat(acc, [col]);
    };

    return columns.reduce(reducer, []);
  }, [hiddenColumns, columns]);
  var toggleColumn = (0, _react.useCallback)(function (id) {
    hiddenColumns.includes(id) ? setHiddenColumns(hiddenColumns.filter(function (col) {
      return col !== id;
    })) : setHiddenColumns([].concat(hiddenColumns, [id]));
  }, [hiddenColumns]);

  var showAllColumns = function showAllColumns() {
    setHiddenColumns([]);
  };

  var hideAllColumns = function hideAllColumns() {
    setHiddenColumns(columns.map(function (col) {
      return col.id;
    }));
  };

  return {
    columns: columns,
    visibleColumns: visibleColumns,
    hiddenColumns: hiddenColumns,
    toggleColumn: toggleColumn,
    showAllColumns: showAllColumns,
    hideAllColumns: hideAllColumns
  };
}
