// @vtex/styleguide/lib/EXPERIMENTAL_useCheckboxTree/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionType = undefined;
exports.default = reducer;

var _util = require("./util");

function reducer(state, action) {
  switch (action.type) {
    case ActionType.Check:
      {
        return [].concat(state, [action.item]);
      }

    case ActionType.Uncheck:
      {
        var _action$itemToToggle = action.itemToToggle,
            item = _action$itemToToggle.item,
            comparator = _action$itemToToggle.comparator;
        return state.filter(function (row) {
          return !comparator(row)(item);
        });
      }

    case ActionType.BulkCheck:
      {
        var itemToToggle = action.itemToToggle;
        if (!itemToToggle) return state;
        return (0, _util.getBulkChecked)(state, itemToToggle.item, itemToToggle.nodesKey, itemToToggle.comparator);
      }

    case ActionType.BulkUncheck:
      {
        var _itemToToggle = action.itemToToggle;
        if (!_itemToToggle) return state;
        return (0, _util.getBulkUnchecked)(state, _itemToToggle.item, _itemToToggle.nodesKey, _itemToToggle.comparator);
      }

    case ActionType.Toggle:
      {
        var _itemToToggle2 = action.itemToToggle;
        if (!_itemToToggle2) return state;
        return (0, _util.getToggledState)(state, _itemToToggle2.item, _itemToToggle2.nodesKey, _itemToToggle2.comparator);
      }

    default:
      {
        return state;
      }
  }
}

var ActionType = exports.ActionType = undefined;

(function (ActionType) {
  ActionType[ActionType["Check"] = 0] = "Check";
  ActionType[ActionType["Uncheck"] = 1] = "Uncheck";
  ActionType[ActionType["Toggle"] = 2] = "Toggle";
  ActionType[ActionType["BulkCheck"] = 3] = "BulkCheck";
  ActionType[ActionType["BulkUncheck"] = 4] = "BulkUncheck";
})(ActionType || (exports.ActionType = ActionType = {}));
