// @vtex/styleguide/lib/EXPERIMENTAL_Table/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _index = require("../Checkbox/index.js");

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Checkbox = function Checkbox(_ref) {
  var id = _ref.id,
      onClick = _ref.onClick,
      props = _objectWithoutPropertiesLoose(_ref, ["id", "onClick"]);

  return _react2.default.createElement("div", {
    onClick: function onClick(e) {
      return e.stopPropagation();
    }
  }, _react2.default.createElement(_index2.default, _extends({
    id: id,
    name: "row_" + id,
    onChange: function onChange() {
      return onClick(id);
    }
  }, props)));
};

Checkbox.defaultProps = {
  id: '',
  onClick: function onClick() {},
  checked: false,
  partial: false,
  disabled: false
};
exports.default = Checkbox;
