// @vtex/styleguide/lib/EXPERIMENTAL_Table/LineActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _ActionMenu = require("../../ActionMenu");

var _ActionMenu2 = _interopRequireDefault(_ActionMenu);

var _OptionsDots = require("../icon/OptionsDots");

var _OptionsDots2 = _interopRequireDefault(_OptionsDots);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var LineAction = function LineAction(_ref) {
  var lineActions = _ref.lineActions,
      rowData = _ref.rowData;
  return _react2.default.createElement(_ActionMenu2.default, {
    buttonProps: {
      variation: 'tertiary',
      icon: _react2.default.createElement(_OptionsDots2.default, null)
    },
    options: lineActions.map(function (action) {
      return _extends({}, action, {
        label: action.renderLabel ? action.renderLabel({
          rowData: rowData
        }) : action.label,
        onClick: function onClick() {
          return action.onClick({
            rowData: rowData
          });
        }
      });
    })
  });
};

exports.default = LineAction;
